### DISCUSSION TITLE
Topology visualization

### DISCUSSION THREAD

**Discussion by SbiCA** [ravendb/ravendb#15384] [Ideas]
Hi @ayende,

Have you thought on a visualizer of ETL, database groups accross clusters? In one of your last webinars I had the feeling you felt the pain as well if the setup involves more than one cluster and/or database group to see what "On going task" is running where and what it's state is.

I was wondering if there would be a way to either extend the current raven studio view? or use something like 
- https://structurizr.com/
- https://mermaid-js.github.io/mermaid/#/
to draw it if there is an API that can provide those kind of insights.

Cheers

**Comment by garayx** [comment]
Hi,

In RavenDB 6 we plan to implement a server-wide studio, so those studio views will be server wide. 

There work currently is in progress: https://issues.hibernatingrhinos.com/issue/RavenDB-17107

**Comment by ayende** [comment]
Note that the problem is fairly complex, because you may have many databases, many ETL processes and multiple clusters.
At that point, the problem is just making sense of a lot of data, which in general is tough.

**Comment by SbiCA** [comment]
@garayx  that will certainly help to understand what happens on one server across database/groups specially for live troubleshooting. 

At least in our scenario this is less about live data or seeing what's happing at the moment (of course would be perfect if that's solved as well) but rather having a big picture about which node/db replicates to which other cluster so that you can make sense of it. I assume sort of graph structure that could be visualized ideally in raven or any of the tools above

@ayende yes I'm totally aware of that but having a similar setup to what you mentioned in the webinar but lots more database and sometimes mixtures of external replication/ETL, filtered replication (which a great features and work really well) in isolation or once you design the solution it all is pretty straightforward due to the knowledge being present but if murphy kicks in 6months (or an arbitrary amount later) at least I have a hard time to remember all the details.  so my concrete problems with current situation is

- complexity grows that grows overtime as described above
- troubleshooting if there is a multitude of ongoing tasks
- onboarding new people 

Also I'm not expecting a perfect solution and would be open to help where I can

**Reply by SbiCA** [reply]
@ayende @garayx  is there already an API exposing the ongoing tasks with the connection attached to it or would that be something that needs to be built first? Because at least IMHO this is the minimum information to get started in creating a graph of cluster, database groups and tasks

**Reply by ayende** [reply]
What do you mean "connection attached" ?

**Reply by SbiCA** [reply]
@ayende  I mean to know where a ongoing task is pointing to e.g. target cluster/database group (what you configure in studio as connection string) . That way it should be possible to draft a first version of a graph probably it needs some help in terms of inputs initially

**Reply by ayende** [reply]
Problem is that you have an agent issue.
You have 3 clusters, two of them talk to one another (replication, etl)
The third is doing ETL to one of them and replication to the other, as well as SQL etl.

Where do you see the global view? What about trust boundaries?

**Reply by SbiCA** [reply]
Generically speaking yes I agree , but since the operator can provide the access to each cluster e.g. 
```
a.eu ... => cert a
a.us ... => cert b
a.ap ... => cert c
```
It should be possible to extract each clusters isolated view and stich it together to one logical view. Similar to what site crawlers do following the links and build up a site map but of course that only works if the information is extractable (hence the question about an endpoint or a way to fetch it)

**Reply by ayende** [reply]
That would be *really* complicated operationally, since you are trying to turn each one of those instances into a reporting dashboard, and the permissions may not always even be there.

@georgiosd suggestion with regards to Grafana is the way to go, I think. You need an _external_ entity to manage that, not the individual components.

**Reply by SbiCA** [reply]
@ayende 
**Agreed** on that it would be complicated if you were to try setting it up within raven/raven studio, but my point is if there is an API available to extract to info so you could build this up _externally_ (hence why I named tools like structirizr or mermaidjs) because at least in my use case(s) that setup is pretty static and it would already help immensely to generate a picture (the 10k foot view and maybe the 100m view). 

see [structurizr sample with dsl](https://structurizr.com/dsl?example=amazon-web-services)
or [mermaidjs flow diagram](https://mermaid.live/edit#pako:eNpVkMFqwzAMhl9F6NRB8wI5DNak7aWwwXpanIOI1dp0tozjMEqSd5-zrtDqJOn__h-hETvRjCWeIwUDx1p5yPXWVCbaPjnqWyiK12nPCZx4vk6wWe0FeiMhWH9-ufGbBYJqPCwYQzLWX-abVP353z1PUDcHCklC-6gcf2SCbWM_TI5_Vkzk7No1JypPVHQUoaLY4hodR0dW57PHxaAwGXassMytpnhRqPycuSFoSrzVNknEnPHd8xppSPJ59R2WKQ58h2pL-QXuvgzkv0T-x_kXgwRfkw)

@georgiosd 
if there is a way to solve this with grafana even better at least as of my current understanding that's not how granafa is inteded to be used because your trying to visualize cluster structures and relations between them. The only plugin I m aware of that could relate to this kind of visualization is this one https://grafana.com/grafana/plugins/agenty-flowcharting-panel/ but to me that sound like a lot of work to be honest

**Comment by georgiosd** [comment]
Could be useful addition to grafana plugin? Overall was thinking it would
be awesome to be able to query RDB internals with RQL, like you can with
SQL system dbs.

On Wed, 21 Dec 2022 at 17:21, Simon Birrer ***@***.***> wrote:

> Generically speaking yes I agree , but since the operator can provide the
> access to each cluster e.g.
>
> a.eu ... => cert aa.us ... => cert b
> a.ap ... => cert c
>
> It should be possible to extract each clusters isolated view and stich it
> together to one logical view. Similar to what site crawlers do following
> the links and build up a site map but of course that only works if the
> information is extractable (hence the question about an endpoint or a way
> to fetch it)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/discussions/15384#discussioncomment-4468626>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMYG5WMQK6DWJLVWUBZ6ZTWOMOBBANCNFSM6AAAAAAST4S3FM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Georgios Diamantopoulos
Founder
Zero to MVP
https://zerotomvp.dev
